ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -228,20 +237,23 @@ private TransactionImple addImportedTransaction(TransactionImple importedTransac
 
 		TransactionImple txn = holder.get();
 
+		// Should only be called by the recovery system - this will replace the Transaction with one from disk
+		if (recoveredTransaction!= null) {
+			synchronized (holder) {
+				// now it's safe to add the imported transaction to the holder
+				recoveredTransaction.recordTransaction();
+				txn = recoveredTransaction;
+				holder.set(txn);
+			}
+		}
+
 		if (txn == null) {
```

Comment:
I would prefer
```java
} else if (txn == null) {
```
here since you have just checked that recoveredTransaction != null